common/achievement/get.ts

```diff
@@ -0,0 +1,214 @@
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID, AND: { template: { isActive: true } } },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = await generateReorderedAchievementData(userAchievementGroups, user);
+    return achievements;
+};
+
+const generateReorderedAchievementData = async (groups: { [group: string]: User_achievement[] }, user: User): Promise<Achievement[]> => {
+    const groupKeys = Object.keys(groups);
+    const achievements = await Promise.all(
+        groupKeys.map(async (key) => {
+            const group = groups[key];
+            const sortedGroupAchievements = group.sort((a, b) => a.groupOrder - b.groupOrder);
+            if (sortedGroupAchievements[0].template.type === achievement_type_enum.TIERED) {
+                return await Promise.all(
+                    sortedGroupAchievements.map(async (groupAchievement) => {
+                        const achievement: Achievement = await assembleAchievementData([groupAchievement], user);
+                        return achievement;
+                    })
+                );
+            }
+            const groupAchievement: Achievement = await assembleAchievementData(sortedGroupAchievements, user);
+            return [groupAchievement];
+        })
+    );
+    return achievements.flat();
+};
+
+const assembleAchievementData = async (userAchievements: User_achievement[], user: User): Promise<Achievement> => {
+    let currentAchievementIndex = userAchievements.findIndex((ua) => !ua.achievedAt);
+    currentAchievementIndex = currentAchievementIndex >= 0 ? currentAchievementIndex : userAchievements.length - 1;
+
+    const achievementContext = transformPrismaJson(user, userAchievements[currentAchievementIndex].context);
+    const currentAchievementTemplate = getCurrentAchievementTemplateWithContext(userAchievements[currentAchievementIndex], achievementContext);
+
+    const achievementTemplates = await prisma.achievement_template.findMany({
+        where: { group: currentAchievementTemplate.group },
+        orderBy: { groupOrder: 'asc' },
+    });
+
+    const state: achievement_state = getAchievementState(userAchievements, currentAchievementIndex);
+
+    const isNewAchievement = state === achievement_state.COMPLETED && !userAchievements[currentAchievementIndex].isSeen;
+
+    const condition = currentAchievementTemplate.condition.includes('recordValue')
+        ? currentAchievementTemplate.condition.replace('recordValue', (userAchievements[currentAchievementIndex].recordValue + 1).toString())
+        : currentAchievementTemplate.condition;
```

Comment:
This is not needed

Reply:
I think it is, why should it not be?

Reply:
Because the condition is only used for `evaluateAchievement`, where we have some logic to be able to use the `recordValue` without replacing it.
Nevertheless, I just noticed that we are  doing 
```
userAchievements[currentAchievementIndex].recordValue + 1
```
here. Why is that?